qualtran/_infra/data_types.py

```diff
@@ -533,34 +534,107 @@ class QFxp(QDType):
         bitsize: The total number of qubits used to represent the integer and
             fractional part combined.
         num_frac: The number of qubits used to represent the fractional part of the real number.
-        signed: Whether the number is signed or not. If signed is true the
-            number of integer bits is reduced by 1.
+        signed: Whether the number is signed or not.
     """
 
     bitsize: SymbolicInt
     num_frac: SymbolicInt
     signed: bool = False
 
+    def __attrs_post_init__(self):
+        if not is_symbolic(self.num_qubits) and self.num_qubits == 1 and self.signed:
+            raise ValueError("num_qubits must be > 1.")
+        if not is_symbolic(self.bitsize) and not is_symbolic(self.num_frac):
+            if self.signed and self.bitsize == self.num_frac:
+                raise ValueError("num_frac must be less than bitsize if the QFxp is signed.")
+            if self.bitsize < self.num_frac:
+                raise ValueError("bitsize must be >= num_frac.")
+
     @property
     def num_qubits(self):
         return self.bitsize
 
     @property
     def num_int(self) -> SymbolicInt:
-        return self.bitsize - self.num_frac - int(self.signed)
+        return self.bitsize - self.num_frac
```

Comment:
This doesn't seem consistent with the docstring anymore?
> In total there are `bitsize = (n_sign + num_int + num_frac)` bits used

Reply:
thanks!